### PR TITLE
[RDY] Disallow rotten components in non-perishable crafts (plus refactoring and infrastructure)

### LINKED ISSUE #22129: Rotten food can be used in non-perishable recipes to make non rotten food.
Game version: current

Operating system: any

Tiles or curses: any

Mods active: none

### Expected behavior
Only foods that allow rotten food should allow rotten food. 

### Actual behavior
Any recipe can be used with rotten food but only the end product is rotten if the end product is perishable. 

### Steps to reproduce the behavior
Get some rotten meat and make dehydrated meat. Then re-hydrate the meat. Its suddenly no longer rotten. 

Suggest creating a check in code that will warn the player if they are going to be using rotten food for the recipe for foods that will produce a rotten result and block foods that would allow the player to exploit this to make non rotten from rotten food. Suggest a Json flag that can be checked if the recipe allows rotten food. The fish bait is a good example where it should be allowed to use rotten food. Maybe ALLOW_ROTTEN or on foods that it shouldn't be allowed a NO_ROTTEN flag.

### PR DESCRIPTION
**Pull request by ifreund**
#### Summary
```SUMMARY: Bugfixes "Disallow rotten components in non-perishable crafts"```

#### Purpose of change
Fixes #22129
Provide infrastructure for further filtering of components based on recipe properties.

#### Describe the solution
Refactor some bad ideas I had about filtering back in #28473.
Pass a `const std::function<bool( const item & )> &filter` pretty much everywhere in crafting.
Add a new function to `recipe` that returns such a filter based on the recipe's properties.

#### Describe alternatives you've considered
Really wish there was a bit of a nicer way to push this stuff into `requirements` but couldn't think of any other options.

#### Additional context
It is intentional that the filter parameter has no default value to hopefully ensure that `recipe::get_component_filter()` is always properly passed.

### CODE REVIEW

**Comment by ifreund** [comment]
jenkins rebuild

File: src/activity_handlers.cpp [hunk 1]
```diff
@@ -2632,8 +2630,24 @@ void activity_handlers::try_sleep_finish( player_activity *act, player *p )
...
+    item *craft = act->targets.front().get_item();
+
+    if( !craft->is_craft() ) {
+        debugmsg( "ACT_CRAFT target '%s' is not a craft.  Aborting ACT_CRAFT.", craft->tname() );
+        p->cancel_activity();
+        return;
+    }
+    if( !p->has_item( *craft ) ) {
+        p->add_msg_if_player( "%s no longer has the target '%s.'  Aborting ACT_CRAFT.",
+                              p->disp_name(), craft->tname() );
```
Comment (codemime) [inline src/activity_handlers.cpp]: This can appear during normal play, so it shouldn't look like a debug message.

**Comment by paulenka-aleh** [comment]
Seems reasonable but does it mean no more seeds from rotten vegetables?

**Comment by ifreund** [comment]
Yes it does, but I think those should be moved to be disassembly recipes anyways. Will get to work on that change.

**Comment by Reclusive-reptile** [comment]
Another thing to consider is the crafting of fertilizer from rotten manure. Crops should prefer their dung to be composted. The crafting of fertilizer should be replaced with a compost heap structure that takes in compostable material and transforms it into fertilizer as said material rots.

**Comment by ifreund** [comment]
Alright, sounds like this may be better handled with a json flag then.  Anyone think of other non-perishables for which an ```ALLOW_ROTTEN``` flag should be used?

**Comment by ifreund** [comment]
Got some failing tests and found some refactoring I want to squeeze in

**Comment by kevingranade** [comment]
I was able to dehydrate rotten meat, making rotten dehydrated meat.
But after re-hydration, it's still rotten.  In a way this addressed #22129, but is that what you intended?

**Comment by ifreund** [comment]
Just to clarify for anyone reading this, dehydrated meat is still perishable, so it was unaffected by this change.  Protein powder, however, can no longer be crafted from rotten dehydrated meat.

**Pull request merged**